File: O73.java
import java.util.Arrays;

public class O73 {
    public int minEatingSpeed(int[] piles, int h) {
        Arrays.sort(piles);
        int left = 1, right = piles[piles.length-1];
        while(left!=right){
            int mid = (left + right) / 2;
            if (isFinish(mid, piles, h))
                right = mid;
            else left = mid+1;
        }
        return left;
    }

    public boolean isFinish(int speed, int[] piles, int h){
        int hour = 0;
        for(int e : piles){
            hour += ((e/speed) + (e%speed==0?0:1));
        }
        return hour<=h;
    }
}

File: O73.py
from math import ceil

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def isFinish(speed) -> bool:
            tempH = 0
            for i in range(len(piles)):
                tempH += ceil(piles[i] / speed)
                if tempH > h:
                    return False
            return True

        left, right = 1, max(piles)
        while not left == right:
            mid = (left + right) // 2
            if isFinish(mid):
                right = mid
            else:
                left = mid+1
        return left

